Updated training and data loading code.

mahmoudnafifi opened the pull request:
Updated training code, input.py, io.py, and ops.py.

Added example of GehlerShi params and hyperparams json files.

Importing tf v2 instead of v1 and do some changes to use tf v2 functions (or compat.v1 functions) accordingly.

Changed the code to import from the ffcc module.

-----------------------------------------

Training process and data loading (train.py, io.py, input.py): The code now follows the same style of the Matlab version by loading the data and ground-truth illuminants from the '../data/project_name' directory. The code can run on trianing/testing style by selecting one of the three folds as a testing set or can do 3-fold cross validation by setting TEST_FOLD = 0. The code loads the hyperparams and params from json if PROJECT_NAME_hyperparams.json and PROJECT_NAME_params.json files exist
TODO: tuning if the json files not exist.

There are some problems in ops.py, where the batch_num is not accessible durint training. This issue is fixed now, except for the compute_chroma_histogram function (TODO).

mahmoudnafifi committed:
Updated training and data loading code.

Updated training code, input.py, io.py, and ops.py.

Added example of GehlerShi params and hyperparams json files.

Importing tf v2 instead of v1 and do some changes to use tf v2 functions (or compat.v1 functions) accordingly.

Changed the code to import from the ffcc module.

-----------------------------------------

Training process and data loading (train.py, io.py, input.py): The code now follows the same style of the Matlab version by loading the data and ground-truth illuminants from the '../data/project_name' directory. The code can run on trianing/testing style by selecting one of the three folds as a testing set or can do 3-fold cross validation by setting TEST_FOLD = 0. The code loads the hyperparams and params from json if PROJECT_NAME_hyperparams.json and PROJECT_NAME_params.json files exist
TODO: tuning if the json files not exist.

There are some problems in ops.py, where the batch_num is not accessible durint training. This issue is fixed now, except for the compute_chroma_histogram function (TODO).

fbleibel-g commented on python/GehlerShi_hyperparams.json:
```diff
@@ -0,0 +1 @@
+{"learning_rate": 0.1, "mult_bias_l2": 0.9786590347278084, "mult_bias_tv": 82.62230380323528, "mult_filters_l2": [82.82293408957314, 0.01], "mult_filters_tv": [44.455099512660816, 0.01], "batch_size": 32, "mult_smooth": 1e-06}
```
(file comment) could we put this file in a 'projects' folder?

fbleibel-g commented on python/ffcc/ellipse.py:
```diff
@@ -119,7 +119,7 @@ def project(x, w_mat, b_vec):
 
   d = distance(x, w_mat, b_vec)
-  scale = tf.rsqrt(tf.maximum(d, 1))
+  scale = tf.compat.v1.rsqrt(tf.maximum(d, 1))
```
If tf.rsqrt doesn't exist anymore in tf2 (have you checked?), I think it's available at tf.math.rsqrt. Let's use that instead of the compat version>

fbleibel-g commented on python/ffcc/fft.py:
```diff
@@ -151,8 +153,8 @@ def vec_to_fft2(v):
     f_complex_right = tf.concat([
-        tf.reverse(tf.conj(f_complex_left[:, 0:1, 1:s, :]), axis=[2]),
-        tf.reverse(tf.conj(f_complex_left[:, 1:, 1:s, :]), axis=[1, 2])
+        tf.reverse(tf.compat.v1.conj(f_complex_left[:, 0:1, 1:s, :]), axis=[2]),
```
Same remark - can we use the non-compat version of conj ?

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -14,145 +14,153 @@
+      'scene_name_weight.txt' does not exist, it will be set to 1.
+    rgb: RGB image (float32)
+    extended_feature: value of the extended feature for this sample (
+      float32). If not 'scene_name_extended_feature.txt' does not exist,
```
typo: remove 'not' (-> does not exist). Here and at line 38.

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -14,145 +14,153 @@
+  rgb = tf.io.decode_image(tf.io.read_file(
+    os.path.join(dir_path, scene_name + '.png')), dtype=tf.dtypes.uint16)
+  normalized = tf.cast(rgb, dtype=tf.float32) / tf.uint16.max
+  tf.compat.v1.disable_eager_execution()
```
disable_eager_execution is a dangerous and broad statement which applies to the entire program. It's usually specified at the beginning of the main python script. I am hoping that we won't need to do that and choose between tf v1 and v2 here.

If you can't use normalized.eval without disable_eager_execution, can you maybe try to evaluate in a small session, e.g. in this toy example:

```
sess = tf.compat.v1.Session()

with sess:
  a = tf.constant(5)
  b = tf.constant(5)

  c = a + b
  value = sess.run([c])
print(value)
```

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -14,145 +14,153 @@
-            extended_feature.txt                # extended feature value.
-          illuminant.txt                        # color of the illuminant in RGB
-          weight.txt                            # weighting
+    scene_name: scene name (e.g., '000001')
```
This is used as a file prefix; can we name this 'scene_prefix'?

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -14,145 +14,153 @@
-    output_dir: The output folder. The record will be saved as <name>.tfrecord.
+    files: A list of filenames that should be synched with the cvfolds.
+    cvfolds: A list of corresponding fold number of each filename in 'files'
+    test_fold: Number of testing fold (should be in the range [1-3]).
```
Note: the comment should read 'index of testing fold'.

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -14,145 +14,153 @@
+  train_folds = list({1, 2, 3} - {test_fold})
+  train_set = []
+  for train_fold_i in train_folds:
+    train_set = np.append(train_set,
```
This is just a python list - can you use list.append instead?

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -14,145 +14,153 @@
+      filenames, respectively.
   """
+  assert len(files) == len(cvfolds)
+  assert test_fold in range(1,4)
```
I think this can easily be made more generic by passing in the number of expected folds to this function and making it a project parameter. Or just to keep it simple, hardcode the number of folds and check as:

assert test_fold in range(1, NUM_CROSS_VALIDATION_FOLDS + 1)

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -14,145 +14,153 @@
-    data: A dictionary maps to the FFCC data bundle.
-    output_dir: The output folder. The record will be saved as <name>.tfrecord.
+    files: A list of filenames that should be synched with the cvfolds.
+    cvfolds: A list of corresponding fold number of each filename in 'files'
```
Should we name cvfolds 'cv_fold_index' instead? Also mention cross validation in the comment, to explain where the 'cv' comes from.

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -14,145 +14,153 @@
+    extended_feature_set.append(extended_feature)
+
+  if shuffle:
+    # Fixed the random seed.
```
let's fix this typo: "Fix the random seed"

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -14,145 +14,153 @@
+  """Encode training data into input/label dictionaries.
+
+  Args:
+    files: full path for scene names
```
What's the type of files?

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -14,145 +14,153 @@
-
-def decode_from_tfrecord_proto(example_proto):
-  """Decodes TFRecord protobuf.
+  eval_set = np.array(files)[(cvfolds == test_fold)]
```
I think the python list comprehension is slightly easier to read:

eval_set = [f for i, f in enumerate(files) if cvfolds[i] == test_fold]

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -14,145 +14,153 @@
+  weight_set = []
+  rgb_set = []
+  extended_feature_set = []
+  for f in iter(files):
```
if 'files' is a list, no need to use iter(files).

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -172,26 +180,39 @@ def _list_subfolders(path):
+    cvfolds_file.seek(cvfolds_file.tell() - 1, os.SEEK_SET)
+    cvfolds_file.truncate()
+    cvfolds_file.close()
+
```
What does this sequence of file operations achieve? If you just want to write to the file in append mode, pass 'a' in open?

Also use with: with open(...) as cvfolds_file: ...

mahmoudnafifi replied:
This is to avoid the extra newline added after the final element. I will use with open(..).

fbleibel-g replied:
We shouldn't need to avoid the extra newline - the parsing code should be robust to newlines. If you read a file with (for line in file: ...), I think this excludes the blank space at the end. Can you point me to the code that reads this data?

mahmoudnafifi replied:
Honestly, I didn't try it without this handling, so yes it may work without a need for this truncate operation. Here is the loading line:
https://github.com/mahmoudnafifi/ffcc/blob/fc83895bf1c39f1ac3f58a347143caa778195023/python/ffcc/io.py#L196

fbleibel-g replied:
Yes, let's try the simple solution and see if that works.

mahmoudnafifi replied:
It is handled automatically by the parsing code. Fixed!

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -172,26 +180,39 @@ def _list_subfolders(path):
+    for fold in range(3):
+      cvfolds[indices[np.floor((fold * number_of_files) / 3).astype(int):
+                      np.floor(((fold + 1) * number_of_files) / 3).astype(int)
+              ]] = int(fold + 1)
```
- This line is difficult to parse. Can you break down the indices[...] into a variable to index cvfolds by?
- You can condense np.floor(x / 3).asint() by using integer division directly: x // 3.
- int(fold + 1) -> you don't need 'int'; fold is already an integer.

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -172,26 +180,39 @@ def _list_subfolders(path):
+  if tf.io.gfile.exists(os.path.join(path, 'cvfolds.txt')):
+    cvfolds = np.loadtxt(os.path.join(path, 'cvfolds.txt'), dtype=np.int)
+  else:
+    # create a 3-fold cross-validation partition and store it in the root path
```
Let's log that this happens (e.g. using tf.logging).

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -172,26 +180,39 @@ def _list_subfolders(path):
+
+  eval_input, eval_label = build_dataset_dict(eval_files, shuffle)
+
+  return training_input, training_label, eval_input, eval_label
```
Note the 'no newline at end of file' warning symbol in github; you should add newlines at the end of each file. I think you should be able to configure your editor to enforce this for you.

fbleibel-g commented on python/train.py:
```diff
@@ -18,61 +18,70 @@
 
-# This directory stores the TFRecord for training.
-DATA_DIR = ''
+os.environ['CUDA_VISIBLE_DEVICES'] = '-1'
```
This new line warrants a comment. What are we doing this for? Probably best not to clobber os.environ, which is global to your entire python process if you can avoid it, although it's fine if there's no other way to achieve this.

fbleibel-g commented on python/train.py:
```diff
@@ -18,61 +18,70 @@
 
 # Model dir for checkpoints, etc.
-MODEL_DIR = ''
+MODEL_DIR = './GehlerShi_model/'
```
Can this be just 'GehlerShi_model', since it's relative to the current directory by default?

fbleibel-g commented on python/train.py:
```diff
@@ -18,61 +18,70 @@
 NUM_EPOCHS = 150
 
+# Number of testing fold in the 3-fold cross validation
+TEST_FOLD = 1 # if TEST_FOLD=0, the code will train three models, each of
```
Comments should go before the code they comment on (or at the end of the line, but not continuing to a new line in that case).

fbleibel-g commented on python/train.py:
```diff
@@ -18,61 +18,70 @@
 
+# Number of testing fold in the 3-fold cross validation
+TEST_FOLD = 1 # if TEST_FOLD=0, the code will train three models, each of
+# which  will be traind with two different folds.
```
typo : trained

fbleibel-g commented on python/train.py:
```diff
@@ -18,61 +18,70 @@
 # The number of epochs for training.
 NUM_EPOCHS = 150
 
+# Number of testing fold in the 3-fold cross validation
```
'index' of testing fold. And it's not obvious that the index starts as 1 in python; you should make that clear(er) in the comments.

fbleibel-g commented on python/train.py:
```diff
@@ -18,61 +18,70 @@
 NUM_UPDATES_PER_EPOCH = 10
 
+# To print the eval results on the latest checkpoint.
+do_print_eval = False
```
If you intend to keep this, name this as a module constant: DO_PRINT_EVAL

fbleibel-g commented on python/train.py:
```diff
@@ -18,61 +18,70 @@
-    train_set: Name of the tfrecords to load as part of the train set.
-    eval_set: Name of the tfrecords to load as part of the eval set.
+    train_set: A list of training data dictionaries generated by
+    'io.read_dataset_from_files'.
```
Google python style guide: indent continuation by 2 characters
" If the description is too long to fit on a single 80-character line, use a hanging indent of 2 or 4 spaces more than the parameter name "

fbleibel-g commented on python/train.py:
```diff
@@ -115,56 +125,107 @@ def print_eval(params, hparams):
+    with open(PROJECT_NAME + '_hyperparams.json', 'r') as hparams_file:
+      hparams = json.load(hparams_file)
+  else:
+    # TODO(fbleibel): populate params and hparams based on project files.
```
I think you addressed this TODO - let's remove this line.

fbleibel-g commented on python/train.py:
```diff
@@ -115,56 +125,107 @@ def print_eval(params, hparams):
+    print('Loading hyperparameters from %s_params.json' % PROJECT_NAME)
+    with open(PROJECT_NAME + '_params.json', 'r') as params_file:
+      params = json.load(params_file)
+  else:
```
Not sure if it's helpful to define a default set of parameters; it's simple to copy them from another project on disk to start a new project. I'd just error loudly (with an assert) if you can't find the params file.

fbleibel-g reviewed (commented):
Thank you so much for making the code compatible with tensorflow 2 - this will make a future migration much easier!

Let's address the histograms in a separate commit, as you suggest.

Mostly looks good; some comments regarding python style formatting and some nits.

mahmoudnafifi committed:
Updated training and data loading code.

Updated training code, input.py, io.py, and ops.py.

Added example of GehlerShi params and hyperparams json files in the 'projects/GehlerShi' directory.

Importing tf v2 instead of v1 and do some changes to use tf v2 functions (or compat.v1 functions) accordingly.

Changed the code to import from the ffcc module.

Training process and data loading (train.py, io.py, input.py): The code now follows the same style of the Matlab version by loading the data and ground-truth illuminants from the '../data/project_name' directory. The code can run on trianing/testing style by selecting one of the three folds as a testing set or can do 3-fold cross validation by setting TEST_FOLD = 0. The code loads the hyperparams and params from json if PROJECT_NAME_hyperparams.json and PROJECT_NAME_params.json files exist
TODO: tuning if the json files not exist.

There are some problems in ops.py, where the batch_num is not accessible durint training. This issue is fixed now, except for the compute_chroma_histogram function (TODO).

mahmoudnafifi commented:
I handled the comments and moved hyperparams and params files to projects/GehlerShi directory.

mahmoudnafifi committed:
Updated training and data loading code.

Update: changed the way of writing cv-folds indices to a file.

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -14,145 +14,155 @@
+  train_folds = list({1, 2, 3} - {test_fold})
+  train_set = []
+  for train_fold_i in train_folds:
+    train_set = list.append(train_set,
```
This can just be train_set.append (I was referring to the type, 'list').

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -14,145 +14,155 @@
+  for train_fold_i in train_folds:
+    train_set = list.append(train_set,
+                          np.array(files).transpose()[
+                            (cv_fold_index == train_fold_i)])
```
Change to a regular python list comprehension? I dont't think we need numpy here.

mahmoudnafifi replied:
the problem I faced is that cv_fold_index is a numpy array and cv_fold_index == train_fold_i will give a logical array. Lists need either integer or slice. Any suggestions?

mahmoudnafifi replied:
Fixed!

fbleibel-g commented on python/ffcc/io.py:
```diff
@@ -172,26 +182,36 @@ def _list_subfolders(path):
+  else:
+    # create a 3-fold cross-validation partition and store it in the root path
+    tf.compat.v1.logging.info(
+      'Creating a 3-fold cross-validation partitions and storing their '
```
Remove 'a'

fbleibel-g commented on python/train.py:
```diff
@@ -100,71 +107,105 @@ def train_and_eval_fn(params, hparams, train_set, eval_set):
+    params = json.load(params_file)
+
+  if TEST_FOLD == 0:
+    for test_fold in range(1, 4):
```
Should this refer to your new constant for the number of folds?

fbleibel-g reviewed (commented):
Almost there, some lingering comments.

mahmoudnafifi committed:
Updated training and data loading code

Use io.NUM_CROSS_VALIDATION_FOLDS in train.py, fixed some typos and use list instead of np arrays.

mahmoudnafifi committed:
Changed the name of the variable num_of_folds to num_fold in io.py

mahmoudnafifi committed:
Changed the name of the variable num_of_folds to num_folds in io.py

mahmoudnafifi committed:
Merge branch 'ffcc_py_tf' of https://github.com/mahmoudnafifi/ffcc into ffcc_py_tf

mahmoudnafifi committed:
Added NUM_CORSS_VALIDATION_FOLDS to train.py

Added NUM_CORSS_VALIDATION_FOLDS to train.py and updated the call of io.read_dataset_from_dir functions accordingly.

fbleibel-g reviewed (approved):
Fantastic, thank yoU!

XiaoyuShi97 commented:
@mahmoudnafifi Hi, thanks for providing a tensorflow alternative for this alsome project. I wonder if the tensorflow codes is finished or not?

jonbarron commented:
No, this codebase is not finished. If you're looking for a working implementation, you'll need to use the Matlab version for now.

XiaoyuShi97 commented:
Thanks a lot for your prompt reply.

XiaoyuShi97 commented:
@jonbarron Hi, can you provide an estimated date about when the tensorflow version can come out?

jonbarron commented:
Regrettably, I can't. It definitely won't be released in the immediate future, and it's possible that it won't ever be released.

btwbtm commented:
Hi, can you open the issue bar since the readme file is not sufficient for us to survive.
For example, for my understanding, I should run "Visualize("GehlerShi")". But it seems I should set an additional parameter *params.TRAINING.CROSSVALIDATION_DATA_FOLDER* in PrivateConstants.m. There is no hint on such settings.
Thx

jonbarron commented:
"Issues" are now open. You are welcome to submit PRs that you think would
improve the clarity of the existing code and documentation.

On Sat, Jan 2, 2021 at 1:22 AM btwbtm <notifications@github.com> wrote:

> Hi, can you open the issue bar since the readme file is not sufficient for
> us to survive.
> For example, for my understanding, I should run "Visualize("GehlerShi")".
> But it seems I should set an additional parameter
> *params.TRAINING.CROSSVALIDATION_DATA_FOLDER* in PrivateConstants.m.
> There is no hint on such settings.
> Thx
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/ffcc/pull/10#issuecomment-753451423>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGZFNSCSPS2FCZHDMGTEHLSX3QW3ANCNFSM4Q6USTOA>
> .
>
